### PR TITLE
Add support for local SPM packages. Closes #436

### LINKED ISSUE #436: Analysis across local swiftpm packages in xcodeproj
Can periphery support scanning across local SwiftPM Packages?

### Test case

Example Project: [PeripheryLocalPackage.zip](https://github.com/peripheryapp/periphery/files/7636763/PeripheryLocalPackage.zip)

This project contains an xcodeproj with a mac app, which imports a local swiftPM package, and a couple of used & unused functions for testing periphery output.

`periphery scan --project PeripheryLocalPackage.xcodeproj --targets PeripheryLocalPackage --schemes PeripheryLocalPackage`

`* No unused code detected.`.

I was hoping to have it detect the two unused functions in `LocalPackage.swift`.

### Scanning the individual package

`cd LocalPackage; periphery scan;`

```
Sources/LocalPackage/LocalPackage.swift:5:13: warning: Function 'unusedPublicFunction()' is unused
Sources/LocalPackage/LocalPackage.swift:9:6: warning: Function 'unusedInternalFunction()' is unused
Sources/LocalPackage/LocalPackage.swift:13:13: warning: Function 'usedPublicFunction()' is unused
Sources/LocalPackage/LocalPackage.swift:18:6: warning: Function 'usedInternalFunction()' is unused
```

`cd LocalPackage; periphery scan --retain-public;`

```
Sources/LocalPackage/LocalPackage.swift:9:6: warning: Function 'unusedInternalFunction()' is unused
```

Scanning the individual packages is insufficient as while we can `--retain-public` to see the unused internal function, it cannot tell us if a public member is unused from the main target (or other modules), which is what we'd like to know.

### Desired outcome from original test case:

```
LocalPackage/Sources/LocalPackage/LocalPackage.swift:5:13: warning: Function 'unusedPublicFunction()' is unused
LocalPackage/Sources/LocalPackage/LocalPackage.swift:9:6: warning: Function 'unusedInternalFunction()' is unused
```

### PR DESCRIPTION
**Pull request by ileitch**

### CODE REVIEW

File: Sources/XcodeSupport/XcodeProject.swift [hunk 1]
```diff
@@ -63,8 +64,34 @@ final class XcodeProject: XcodeProjectlike {
...
+
+        if shouldLoadPackages {
+            // The project file does not provide a clear way to identify file references for local SPM packages.
+            // We need to iterate over all references and check for folders containing a Package.swift file.
+            for fileRef in xcodeProject.pbxproj.fileReferences {
+                // To avoid checking every single file reference, narrow our search down to the known file types Xcode uses
+                // for package references.
+                let lastKnownFileType = fileRef.lastKnownFileType
+
+                guard (lastKnownFileType == "text" || lastKnownFileType == "wrapper"),
```
Comment (mthole) [inline Sources/XcodeSupport/XcodeProject.swift:80]: In testing this on my project, I needed to add a case for `lastKnownFileType == "wrapper"` here.

File: Sources/XcodeSupport/XcodeProjectDriver.swift [hunk 2]
```diff
@@ -19,8 +19,33 @@ public final class XcodeProjectDriver {
...
+        var configurationTargets = Set(configuration.targets)
+        var invalidTargetNames: [String] = []
+
+        let targets = project.targets.filter {
+            if configuration.targets.contains($0.name) {
+                configurationTargets.remove($0.name)
+                return true
+            }
+
+            invalidTargetNames.append($0.name)
```
Comment (mthole) [inline Sources/XcodeSupport/XcodeProjectDriver.swift:31]: In my testing, this causes all targets in my .xcodeproj that are not passed to `periphery` to be considered invalid targets.

**Comment by vibrazy** [comment]
I've tested locally and it seems to work with Packages that don't have sub packages inside.
``` 74AE900427C92543002087C4 /* SPMFramework */ = {isa = PBXFileReference; lastKnownFileType = wrapper; path = SPMFramework; sourceTree = "<group>"; };```
However it doesn't work if I have a RootPackage which contains multiple packages inside.
```748B7D7827CCE4A600C77B2B /* RootPackage */ = {isa = PBXFileReference; lastKnownFileType = wrapper; path = RootPackage; sourceTree = "<group>"; };```
Contains InnerFrameworkA and InnerFrameworkB.

**Target 'InnerFrameworkA' does not exist in 'PeripheryTestingWorkspace.xcworkspace'.**
Not sure if this is something you or XcodeProj have the issue with. 
Cheers

**Comment by jorgealegre** [comment]
I'd love to see this merged in

**Comment by oriyentel** [comment]
would absolutely love for this to be merged in as well!

**Comment by brandFromNSK** [comment]
Some news? Can we try it with specific branch

**Comment by ileitch** [comment]
I still plan to finish this, but I'm focusing solely on bug fixes with the limited time I have at the moment.

**Comment by ileitch** [comment]
1 year later, and it's finally merged 😅

**Comment by mjarvis** [comment]
@ileitch Thanks so much! Looking forward to using this functionality!!!!

**Pull request merged**